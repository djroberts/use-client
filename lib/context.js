"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const React = require("react");
exports.ClientRequestContext = React.createContext({});
