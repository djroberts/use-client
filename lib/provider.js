"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const React = require("react");
const context_1 = require("./context");
exports.ClientProvider = (props) => {
    const requestsRef = React.useRef({});
    const { children } = props;
    return React.createElement(context_1.ClientRequestContext.Provider, { value: requestsRef.current }, children);
};
