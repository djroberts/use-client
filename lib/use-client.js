"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const React = require("react");
const reducer_1 = require("./reducer");
const context_1 = require("./context");
const initialState = {
    isLoading: false,
    data: null,
    error: null,
    response: null,
    index: 1,
    statusCode: null,
};
const initialOptions = {
    priority: 'last',
};
exports.useClient = (name, query, options = {}) => {
    const runningOptions = Object.assign(Object.assign({}, initialOptions), options);
    const requests = React.useContext(context_1.ClientRequestContext);
    if (requests[name]) {
        requests[name].call = query;
        if (runningOptions.priority === 'last') {
            requests[name].running = false;
        }
    }
    else {
        requests[name] = {
            call: query,
            running: false,
            index: 0,
        };
    }
    const [state, dispatch] = React.useReducer(reducer_1.reducer, initialState);
    const handleRequest = async (data) => {
        if (runningOptions.priority === 'first' && requests[name].running) {
            return;
        }
        requests[name].index = state.index;
        dispatch({
            type: 'start',
        });
        try {
            const promise = query.call(data);
            requests[name].running = true;
            const response = await promise;
            if (state.index !== requests[name].index) {
                return;
            }
            dispatch({
                type: 'success',
                response,
            });
            requests[name].running = false;
        }
        catch (error) {
            dispatch({
                type: 'success',
                error,
            });
            requests[name].running = false;
        }
    };
    return Object.assign(Object.assign({}, state), { handleRequest });
};
