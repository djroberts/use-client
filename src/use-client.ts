import * as React from 'react';
import { reducer } from './reducer';
import { ClientRequestContext } from './context';

export interface ClientError {
    [key: string]: any;
}

export interface ClientResponse {
    data: any;
    status: number | string | null;
    [key: string]: any;
}

export interface ClientRequestPromise<T = any> extends Promise<ClientResponse> {}

export type ClientRequestCall = (data?: any) => ClientRequestPromise;

export interface ClientRequest {
    call: ClientRequestCall;
    running: boolean;
    index: number;
    data?: any;
}

export interface ClientRequests {
    [key: string]: ClientRequest;
}

export interface State<T> {
    isLoading: boolean;
    data: T | null;
    error: ClientError | null;
    response: ClientResponse | null;
    index: number;
    statusCode: string | number | null;
}

export interface UseClientResult<T> extends State<T> {
    handleRequest: (data?: any) => void;
    setData: (data: any) => void;
    data: T | null;
}

const initialState: State<null> = {
    isLoading: false,
    data: null,
    error: null,
    response: null,
    index: 1,
    statusCode: null,
};

export interface Action {
    type: 'start' | 'success' | 'error' | 'setData';
    data?: any;
    response?: ClientResponse;
    error?: ClientError;
}

export interface Options {
    priority?: 'first' | 'last';
    cache?: boolean;
}

const initialOptions: Options = {
    priority: 'last',
    cache: false,
};

export const useClient = <T>(name: string, query: ClientRequestCall, options: Options = {}): UseClientResult<T> => {
    const runningOptions: Options = {
        ...initialOptions,
        ...options,
    };

    const requests = React.useContext<ClientRequests>(ClientRequestContext);

    if (requests[name]) {
        requests[name].call = query;
        if (runningOptions.priority === 'last') {
            requests[name].running = false;
        }
    } else {
        requests[name] = {
            call: query,
            running: false,
            index: 0,
        };
    }

    const [state, dispatch] = React.useReducer(reducer, initialState);
    const handleRequest = async (data?: any) => {
        if (runningOptions.priority === 'first' && requests[name].running) {
            return;
        }

        requests[name].index = state.index;

        dispatch({
            type: 'start',
        });

        try {
            const promise = query(data);

            requests[name].running = true;

            const response: ClientResponse = await promise;

            if (state.index !== requests[name].index) {
                return;
            }

            dispatch({
                type: 'success',
                response,
            });

            if (runningOptions.cache) {
                requests[name].data = response.data;
            }

            requests[name].running = false;
        } catch (error) {
            dispatch({
                type: 'success',
                error,
            });

            requests[name].running = false;
        }
    };

    const getData = (): any => {
        if (state.data) {
            return state.data;
        }

        if (!runningOptions.cache) {
            return null;
        }

        if (requests[name]) {
            return requests[name].data;
        }

        return null;
    };

    const setData = (data: any): void => {
        dispatch({
            type: 'setData',
            data,
        });
    };

    return {
        ...state,
        data: getData(),
        setData,
        handleRequest,
    };
};
