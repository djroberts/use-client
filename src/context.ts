import * as React from 'react';

import { ClientRequests } from './use-client';

export const ClientRequestContext = React.createContext<ClientRequests>({});
