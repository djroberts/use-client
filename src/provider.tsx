import * as React from 'react';

import { ClientRequestContext } from './context';
import { ClientRequests } from './use-client';

export interface ClientProviderProps {
    children: React.ReactNode;
}

export const ClientProvider = (props: ClientProviderProps) => {
    const requestsRef = React.useRef<ClientRequests>({});

    const { children } = props;

    return <ClientRequestContext.Provider value={requestsRef.current}>{children}</ClientRequestContext.Provider>;
};
