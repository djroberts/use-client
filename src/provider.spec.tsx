import * as React from 'react'

import { render } from '@testing-library/react'

import { ClientProvider } from './provider'
import { ClientRequestContext } from './context'
import { ClientRequests } from './use-client'

describe('Provider', () => {
    it('it should render without crashing', () => {
        const { container } = render(
          <ClientProvider>
            <div>Child element</div>
          </ClientProvider>,
            {
                container: document.createElement('div'),
            },
        )

        expect(container).toMatchSnapshot()
    })

    it('it should render with an empty context object', async () => {
        expect.assertions(1)

        const promise = new Promise(resolve => {
            const Child = () => {
                const requests = React.useContext<ClientRequests>(ClientRequestContext)

                resolve(requests)

                return <div>Child element</div>
            }

            render(
              <ClientProvider>
                <Child />
              </ClientProvider>,
                {
                    container: document.createElement('div'),
                },
            )
        })

        await expect(promise).resolves.toMatchObject({})
    })
})
