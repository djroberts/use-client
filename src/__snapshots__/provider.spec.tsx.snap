// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Provider it should render without crashing 1`] = `
<div>
  <div>
    Child element
  </div>
</div>
`;
